source/server/listener_manager_impl.cc

```diff
@@ -122,27 +122,45 @@ class ListenerSocketOption : public Network::Socket::Option, Logger::Loggable<Lo
     // This has the side effect of setting the option twice after the socket is first created,
     // both before and after bind().
     UNREFERENCED_PARAMETER(state);
-    int transparent = transparent_;
 
+    const auto* ip = socket.localAddress()->ip();
+    if (!ip) {
+      return false;
+    }
 #ifdef SOL_IP
-    int rc = setsockopt(socket.fd(), SOL_IP, IP_TRANSPARENT, &transparent, sizeof(transparent));
+    int rc;
+    int transparent = transparent_;
+
+    if (ip->version() == Network::Address::IpVersion::v4) {
+      rc = setsockopt(socket.fd(), SOL_IP, IP_TRANSPARENT, &transparent, sizeof(transparent));
+    } else {
+#if defined(SOL_IPV6) && defined(IPV6_TRANSPARENT)
```

Comment:
This won't work for ipv6 is SOL_IP (ipv4) isn't defined. It's also pretty hard to read and understand the nested #ifdefs. 

Can you restructure slightly so that ipv4 and ipv6 paths are independent (ie neither requires the other)?

Reply:
Done.